processing/src/main/java/io/druid/segment/DimensionHandler.java

```diff
@@ -0,0 +1,173 @@
+ */
+public interface DimensionHandler<EncodedType extends Comparable<EncodedType>, ActualType extends Comparable<ActualType>>
+{
+  /**
+   * Creates a new DimensionIndexer, a per-dimension object responsible for processing ingested rows in-memory, used by the
+   * IncrementalIndex. See {@link DimensionIndexer} interface for more information.
+   *
+   * @param lock An object to synchronize on. The IncrementalIndex will pass in its dimensionsDescs
+   *             Map object, which stores descriptors for each dimension.
+   *
+   * @return A new DimensionIndexer object that synchronizes on the provided lock.
+   */
+  public DimensionIndexer makeIndexer(Object lock);
+
+
+  /**
+   * Creates a new DimensionMerger, a per-dimension object responsible for merging indexes/row data across segments
+   * and building the on-disk representation of a dimension.
+   *
+   * See {@link DimensionMerger} interface for more information.
+   *
+   * @param indexSpec     Specification object for the index merge
+   * @param outDir        Location to store files generated by the merging process
+   * @param ioPeon        ioPeon object passed in by IndexMerger, manages files created by the merging process
+   * @param v9Smoosher    FileSmoosher used for combining files created by the merging process.
+   * @param mapper        Object mapper used for JSON serde
+   * @param capabilities  The ColumnCapabilities of the dimension represented by this DimensionHandler
+   * @param progress      ProgressIndicator used by the merging process
+
+   * @return A new DimensionMerger object.
+   */
+  public DimensionMerger makeMerger(
+      IndexSpec indexSpec,
+      File outDir,
+      IOPeon ioPeon,
+      FileSmoosher v9Smoosher,
+      ObjectMapper mapper,
+      ColumnCapabilities capabilities,
+      ProgressIndicator progress
+  );
+
+
+  /**
+   * Creates a new DimensionColumnReader, a per-dimension object responsible for reading data from an on-disk column.
+   * Used by QueryableIndex and associated wrapper classes.
+   *
+   * See {@link DimensionColumnReader} interface for more information.
+   *
+   * @param column The Column object associated with this DimensionHandler's dimension.
+
+   * @return A new DimensionColumnReader object.
+   */
+  public DimensionColumnReader makeColumnReader(Column column);
+
+
+  /**
+   * Creates a DimensionSelector of a type suitable for reading values from this dimension, used for querying and filtering.
+   *
+   * @param cursor        Cursor object created by a StorageAdapter.
+   * @param dimSpec       Specifies extraction function and output name for a dimension.
+   * @param capabilities  The capabilities associated with this handler's dimension.
+   *
+   * @return A new DimensionSelector object linked with the provided cursor and dimSpec
+   */
+  public DimensionSelector getDimensionSelector(ColumnSelectorFactory cursor, DimensionSpec dimSpec, ColumnCapabilities capabilities);
```

Comment:
ColumnCapabilities is unnecessary, I'll take that out, but DimensionSpec is provided by queries, so the handler wouldn't know about it beforehand, and it would change across different calls